lib/twom.c

```diff
@@ -0,0 +1,3294 @@
+                r = TWOM_LOCKED;
+                goto done;
+            }
+            if (errno == EINTR) continue;
+            db->error("read_lock datalock fcntl failed",
+                      "filename=<%s>", db->fname);
+            r = TWOM_IOERROR;
+            goto done;
+        }
+        file->has_datalock = 1;
+    }
+
+    // release the head section (so writers don't starve)
+    fl.l_type= F_UNLCK;
+    fl.l_start = 0;
+    fl.l_len = HEADER_MAGIC_SIZE;
+    for (;file->has_headlock;) {
+        if (fcntl(file->fd, F_SETLKW, &fl) < 0) {
+            if (errno == EINTR) continue;
+            db->error("read_lock headunlock fcntl failed",
+                      "filename=<%s>", db->fname);
+            r = TWOM_IOERROR;
+            goto done;
+        }
+        file->has_headlock = 0;
+    }
+
+    // if the existing map it too small, replace it
+    if (file->size < (size_t)sbuf.st_size) {
+        /* map the new space (note: we map READ|WRITE even for readonly locks,
+         * if we might lock for write later and want to reuse the mmap */
+        if (file->size) munmap(file->base, file->size);
+        int flags = db->readonly ? PROT_READ : PROT_READ|PROT_WRITE;
+        file->base = (char *)mmap((caddr_t)0, sbuf.st_size, flags, MAP_SHARED, file->fd, 0L);
+        file->size = sbuf.st_size;
+        if (!file->base) {
+            db->error("read_lock mmap failed",
+                      "filename=<%s>", db->fname);
+            r = TWOM_IOERROR;
+            goto done;
+        }
+    }
+
+    // reread header
+    r = read_header(db, file, &file->header);
+    if (r) goto done;
+
+    file->committed_size = file->header.current_size;
+    file->written_size = file->committed_size;
+
+    /* we can't read an unclean database */
+    if (!db_is_clean(db, file)) {
+badfile:
+        /* we have to be able to re-lock safely */
+        if (db->readonly) {
+            r = TWOM_IOERROR;
+            goto done;
+        }
+        /* if we take a write lock, that will repair it */
+        unlock(db, file);
+        // no txn, release the write_lock after repairing if needed
+        r = write_lock(db, NULL, file, flags);
+        if (r) return r;
+        /* if we want a transaction, we'll need to re-lock with the readlock */
+        if (txnp) {
+            goto restart;
+        }
+    }
+
+    if (txnp) {
+        if (!*txnp) *txnp = _newtxn_read(db);
+        else if (!(*txnp)->mvcc) {
+            // if we're not on this file, change file
+            if ((*txnp)->file != file) {
+                (*txnp)->file->refcount--;
+                (*txnp)->file = file;
+                (*txnp)->file->refcount++;
+            }
+            (*txnp)->end = file->written_size;
+        }
+
+        return 0;
+    }
+
+done:
+    unlock(db, file);
+    return r;
+}
+
+/***********************  OPEN AND CLOSE *************************************/
+
+static void errors_to_stderr(const char *msg, const char *fmt, ...)
```

Comment:
Adding here `__attribute__((format(printf, 2, 3)))` avoids warnings on the line  1906 `vfprintf(stderr, fmt, args);`:

> format string is not a string literal
> clang-diagnostic-format-nonliteral

Reply:
added